### ISSUE TITLE
[HttpParser] Handle malformed `ipv6:port` in request line or gracefully send `BadRequest` error code

### ISSUE THREAD

**Issue by MrUbun2** [abhinavsingh/proxy.py#500]
I need to make an http and https compatible proxy (and other types of http methods if there are any) for a project I am working on.

I figured out how to make an http proxy but I am stuck at https and so I found this.

I turned it on and set my proxy settings on my laptop and I get an error right away:
```
Exception while handling connection <socket.socket fd=22, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 8899), raddr=('127.0.0.1', 49332)>
* 16 lines omitted *
    self.set_line_attributes()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/proxy/http/parser.py", line 122, in set_line_attributes
    self.port = 443 if self.url.port is None else self.url.port
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/parse.py", line 174, in port
    raise ValueError(message) from None
ValueError: Port could not be cast to integer value as b'xxxx:xxxx:xxx::xxxx:xxx'
```

Help is appreciated.

**Comment by abhinavsingh** [comment]
Hi @MrUbun2 , could you uncomment `# logger.info(data)` here and paste the raw request received which triggers this issue?  Looks like we are trying to parse an ipv6 as port for some reason.

`proxy/core/connection/connection.py`

<img width="718" alt="Screen Shot 2021-02-28 at 12 59 09 PM" src="https://user-images.githubusercontent.com/126065/109411028-ff608a80-79c4-11eb-9cef-17ce20648e7b.png">

**Comment by abhinavsingh** [comment]
@MrUbun2 Are you still facing this issue.  I too use proxy.py on MacOS and it works right out of box for me.

When you say `https`, are you running `proxy.py` in HTTPS mode when encountering this error?

**Comment by abhinavsingh** [comment]
Encountered this today in production.  We'll look into it:

<img width="766" alt="Screen Shot 2021-11-12 at 4 06 06 AM" src="https://user-images.githubusercontent.com/126065/141379254-178fce59-77c1-4251-a9ac-47d7a68f5d97.png">

This looks like a bug in request line parsing and happening when requests contain IPv6:443 in the request line.

**Comment by abhinavsingh** [comment]
@MrUbun2 Ok can confirm that this only happens for malformed requests.  Example, such malformed requests are of the form `IPv6:Port`, while the valid format is `[IPv6]:Port`

To confirm, see:

```python
>>> # Error scenario
>>> parse.urlsplit('http://::').port
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/abhinavsingh/.pyenv/versions/3.10.0/lib/python3.10/urllib/parse.py", line 178, in port
    raise ValueError(message) from None
ValueError: Port could not be cast to integer value as ':'
>>> 
>>> # Valid scenario
>>> parse.urlsplit('http://[::]').port
```

**Comment by abhinavsingh** [comment]
Having said that, we must still be able to make such malformed requests work.  Converting this into an enhancement now (instead of bug).

**Comment by abhinavsingh** [comment]
As explained above happens when request itself is invalid, however, I added a relaxed rule to even allow such requests to go through.  Following PR must fix this:

- #730

**Comment by abhinavsingh** [comment]
- Fixed in #730
